### ISSUE TITLE
WPF Widgets and Grid control

### ISSUE THREAD

**Issue by charlenni** [Mapsui/Mapsui#299]
Hi Paul,

I'm trying to implement the ScaleBar as Widget for WPF too. But no I have problems placing the text at the correct position. If I use a TextBlock for this, the text isn't placed correct, because you used a Grid as layout element.

Wouldn't it be better using a Canvas as parent layout element for Widgets?

I changed this in my version and if I'm getting ready, publish this.

**Comment by charlenni** [comment]
Ok, added a scale bar Widget with renderers for Skia and WPF to the branch [ScaleBar](https://github.com/charlenni/Mapsui/tree/ScaleBar).

While doing this, I changed the parent for Widgets to Canvas instead of Grid. With this, Widgets could place all over the screen, not only in certain places.

@pauldendulk Perhaps you could take a look on this, if it is ok?

**Comment by pauldendulk** [comment]
Did not not take a look yet, but wanted to mention that the Attribution widgets also contain text to can be placed all over the map. Why is this different?

**Comment by charlenni** [comment]
Than I didn't get it working.

If I understand correct, you could place content in a Grid cell only at Left, Right or Center (vertical/horizontal), so you couldn't place it at any position.

**Comment by charlenni** [comment]
Oh, I forgot. You could use Strech too. That fills the cell.

**Comment by pauldendulk** [comment]
The Left/Right/Center settings determine the origin, there is a Offset (or Margin?) property to determine the position. A specific position usually does not work well when the map is resized.

**Comment by charlenni** [comment]
I interpreted it different. Your Widgets have a PositionX and PositionY. There is although a MarginX and MarginY. So I changed the Horizontal-/VerticalAlignment to Left/Right/Center/Position. If you use one of the first three, you could use Margin for a little distance to the frame. With Position you could set absolut distances from (0/0). Works great on both plattforms. Resizing isn't a problem, because Widgets are redrawn each time when the map changes.

**Comment by charlenni** [comment]
The difference to your HyperlinkWidget is, that this isn't a TextBox, but a TextBox inside of a Border. This Border is filled with a background color. I don't have a Border around the TextBox. And with this, the TextBox could not placed at any position in the Grid.

Perhaps it would be possible to place the ScaleBar inside of a Border and than place this Border inside of the Grid.

But is there any problem to use a Canvas instead of a Grid? A Canvas seems much more flexible than a Grid for this. But today was the first time I did working with WPF :)

**Comment by charlenni** [comment]
Btw. I moved your Hyperlink renderer to HyperlinkWidgetRenderer. I think it would be better to have an extra class as renderer for each Widget type. It contains a static function Draw(), which draws the Widget onto the Canvas.

Or isn't a static function a good way to do this?

**Comment by pauldendulk** [comment]
The margin itself is not the colored border. That is the padding. So you could use the margin to position relative to a corner.

**Comment by pauldendulk** [comment]
Wow, this looks fantastic! A few remarks

- When I pan the map the scalebars also resize. This is an error or do I misunderstand how the scalebars function?
- You use BackColor as what we call Halo for the labels. Maybe you could use the Halo property for this here too.
- Did not look at the Canvas vs Grid.
- Widget touch does not seem to work anymore in wpf.

**Comment by charlenni** [comment]
Thank you.

When you pan the map, than the scalebar changes. This is correct, because of the projection of the map. To bring the map on a rect, parts far in the north and far in the south are bigger. If you pan horizonticaly there is no change.

Only Halo or HaloColor?

I look, what's happening with the touch event.

**Comment by charlenni** [comment]
Widget touch now works again :) Your suggestion was correct. The Envelop hadn't the right coordinates. 

Found some mor bugs :(

**Comment by charlenni** [comment]
@pauldendulk Is it correct, that there isn't an event for a Widget touch? What happens, if you touch a Widget is hardcoded in MapControl?

**Comment by charlenni** [comment]
Ok, Paul, I now added a ZoomInOutWidget. For this I needed the WidgetTouched functionality. Refactored this in all MapControls. Didn'Ät changed anything at the Hyperlink handling, but this is odd. Hyperlinks now have an extra handling.

And there is something more strange. With the zoom buttons, you have a standard handling. If you zoom in, the Viewport.Resolution is multiplied with 0.5, for zoom out it is multiplied by 2. Normally you shouldn't have any problems with zooming. But it seems, that the clicks are not correct. If you zoom two times in, and than out, the map zooms three times in. The first click on out with multiply by 2 is like multiplied with 0.5. Perhaps I missed a refresh of the Viewport? Btw. how to do this?

**Comment by pauldendulk** [comment]
@charlenni in the WPF hyperlink touch was a bug that I fixed this morning. The hit area was way too big. Perhaps this is relevant.

**Comment by pauldendulk** [comment]
Each widget renderer in their own class is an improvement. I like to use static methods a lot but in this case I think I will move towards an interface where it will be possible to inject widgets. It should be possible to add widgets that are not part of the Mapsui code base. i have to think about it, but this is a step in the right direction

**Comment by pauldendulk** [comment]
@charlenni Could you create a pull request?

**Comment by charlenni** [comment]
@pauldendulk Do you want to have a pull request for master or do you want to create a new branch?

**Comment by charlenni** [comment]
No, I don't think, that the Hyperlink problem is the reason. For my it seems, that, if I change Viewport.Resolution, the map isn't always refreshed. Could this done by hand?

**Comment by pauldendulk** [comment]
To master. Will not release immediately.

**Comment by charlenni** [comment]
Argh. It seems, that, if you change Viewport.Resolution, MapControl.Refresh() isn't called.

So I now changed from using Viewport to use Map and call NavigateTo() for zoom operations.

**Comment by pauldendulk** [comment]
@charlenni @radderz You both added System.ValueTuple. Why was that? When I remove it the code compiles and I don't see any immediate runtime errors.

**Comment by charlenni** [comment]
I needed the System.ValueTuple for returning more than one variable. Don't understand, why you don't need this.

**Comment by pauldendulk** [comment]
Tuples are part of C#7. I would think you don't need a separate nuget package anymore.

**Comment by charlenni** [comment]
I thought I updated VS 2017 before starting with the work. Perhaps I had not the newest version.

**Comment by pauldendulk** [comment]
after a clean checkout I also get the build error. Let me investigate.

**Comment by charlenni** [comment]
I found this:

Note that if you don’t see the ValueTuple available in your project, you have to download the System.ValueTuple 4.3.0 NuGet package to your project. You don’t need to do anything if you are using .NET Framework 4.7 or higher, or .NET Standard Library 2.0 or higher.

**Comment by pauldendulk** [comment]
Yes, that's it. I have been using tuples without any required packages but that was in .NET Framework 4.7.

**Comment by pauldendulk** [comment]
I removed the Position from the HorizontalAlignment and used MarginX and MarginX to position the control that was previously using Position and it seems to me that you can do the same thing. In the end you always position relative to some corner. 

![image](https://user-images.githubusercontent.com/963462/35742150-e6a652fe-0839-11e8-9c74-a1af92687cd3.png)

Personally I don't think the Position field did really belong in the aligment enum. So if all functionality can be supported without, i'd prefer that. What function did you see for Position?

**Comment by pauldendulk** [comment]
I see you added a reference to the mapsforge source of you code. Good thing you added that, that's important. It does mean that this part can not be moved to an MIT license without the approval of the copyright holders (mapsforge is LGPL). Perhaps that a solution would be to put it in a separate repository and keep that LGPL for that repository. Since I want to allow widget plugins this might not be such a problem.

**Comment by charlenni** [comment]
Ok, I don't need the Position entries really. I thought, sometime someone came up with an idea for this. It would work with Margin too. And than you could again use Grid, even when I don't know, what the advantage of Grid is. But, as I had said, WPF isn't really  new to me.

Than you should remove it from VerticalAlignment too.

The only code from Mapsforge are the UnitConverters. And they are trivial and transfered from Java to C#. Don't know, what to do with them.

**Comment by pauldendulk** [comment]
This was implemented long ago.
